### PR TITLE
Bump CAPI to v1.4.1

### LINKED ISSUE #3297: Test CAPI with release 1.4 RC
Release candidate for v1.4.0 release of CAPI is out: https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.4.0-rc.0

We should bump the version of CAPI in CAPZ and run all the tests with it to help validate the new release and ensure there are no regressions.

See [the Cluster-API book](https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html) for an overview of relevant changes between Cluster API v1.3 and v1.4 that may need changes in CAPZ.

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Updates the cluster-api dependency to recent release v1.4.1. See https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.4.1

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3297

**Special notes for your reviewer**:

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Bump CAPI to v1.4.1
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -21,8 +21,8 @@ require (
 	github.com/hashicorp/go-retryablehttp v0.7.2
 	github.com/hashicorp/golang-lru v0.5.4
 	github.com/jongio/azidext/go/azidext v0.4.0
-	github.com/onsi/ginkgo/v2 v2.8.4
-	github.com/onsi/gomega v1.27.1
+	github.com/onsi/ginkgo/v2 v2.9.1
```
Comment (nawazkh) [inline go.mod:24]: We would also need to update https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/c0dab22666648ea50e91cd9b73fa0b7113826cd1/Makefile#L108

**Comment by CecileRobertMichon** [comment]
let's run the dual stack test too

/test pull-cluster-api-provider-azure-e2-optional

File: .golangci.yml [hunk 2]
```diff
@@ -2,6 +2,11 @@ run:
   deadline: 10m
   skip-files:
   - 'zz_generated\.(\w*)\.go$'
+  skip-dirs:
```
Comment (mboersma) [inline .golangci.yml:5]: Why are we skipping the linter in these directories now? Otherwise, lgtm–nice work, thanks for sticking with this @willie-yao!

Comment (willie-yao) [reply .golangci.yml:5]: We had to skip them due to `v1alpha3` and `v1alpha4` being deprecated in CAPI, which fails the test CI job. This was added to suppress those deprecation warnings. I will be working on removing those API versions completely for the 1.9 release.

**Review by mboersma — COMMENTED** [review]
/lgtm

File: api/v1beta1/azuremanagedmachinepool_webhook_test.go [hunk 3]
```diff
@@ -547,12 +547,12 @@ func TestAzureManagedMachinePoolUpdatingWebhook(t *testing.T) {
 			name: "Can't update SubnetName with error",
 			new: &AzureManagedMachinePool{
 				Spec: AzureManagedMachinePoolSpec{
-					SubnetName: pointer.StringPtr("my-subnet"),
+					SubnetName: pointer.String("my-subnet"),
```
Comment (CecileRobertMichon) [inline api/v1beta1/azuremanagedmachinepool_webhook_test.go:550]: why all the changes from StringPtr to String?

Comment (mboersma) [reply api/v1beta1/azuremanagedmachinepool_webhook_test.go:550]: They're the same function, but the shorter name is now preferred. I think `pointer.StringPtr` was considered stuttering compared to `pointer.String`.

Comment (CecileRobertMichon) [reply api/v1beta1/azuremanagedmachinepool_webhook_test.go:550]: Got it, thanks!

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve

Thank you @willie-yao!

**Comment by nawazkh** [comment]
This is great work! 🚀 Looks good to me!

**Pull request merged**